src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -2473,21 +2495,23 @@ public bool HasBoxingConversion(TypeSymbol source, TypeSymbol destination, ref H
             if ((source.TypeKind == TypeKind.TypeParameter) &&
                 HasImplicitBoxingTypeParameterConversion((TypeParameterSymbol)source, destination, ref useSiteDiagnostics))
             {
-                return true;
+                return ConversionKind.TypeParameterBoxing;
             }
 
             // The rest of the boxing conversions only operate when going from a value type to a
             // reference type.
             if (!source.IsValueType || !destination.IsReferenceType)
             {
-                return false;
+                return ConversionKind.NoConversion;
```

Comment:
from this point on, we know that source is a value type.  i.e. even if it's a type parameer, it's constrained on "struct".